chacha20_poly1305/src/poly1305.rs

```diff
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Poly1305 one-time message authenticator from RFC8439.
+//!
+//! Heavily inspired by the ["Donna"](https://github.com/floodyberry/poly1305-donna/blob/master/poly1305-donna-32.h) implementation in C
+//! and Loup Vaillant's [Poly1305 design article](https://loup-vaillant.fr/tutorials/poly1305-design).
+
+/// 2^26 for the 26-bit limbs.
+const BITMASK: u32 = 0x03ffffff;
+/// Number is encoded in five 26-bit limbs.
+const CARRY: u32 = 26;
+
+/// Poly1305 authenticator takes a 32-byte one-time key and a message and produces a 16-byte tag.
+///
+/// 64-bit constant time multiplication and addition implementation.
+#[derive(Debug)]
+pub struct Poly1305 {
+    /// r part of the secret key.
+    r: [u32; 5],
+    /// s part of the secret key.
+    s: [u32; 4],
+    /// State used to create tag.
+    acc: [u32; 5],
+    /// Leftovers between adds.
+    leftovers: [u8; 16],
+    /// Track relevant leftover bytes.
+    leftovers_len: usize,
+}
+
+impl Poly1305 {
+    /// Initialize authenticator with a 32-byte one-time secret key.
+    pub fn new(key: [u8; 32]) -> Self {
+        // Taken from Donna. Assigns r to a 26-bit 5-limb number while simultaneously 'clamping' r.
+        let r0 =
+            u32::from_le_bytes(key[0..4].try_into().expect("infalliable conversion")) & 0x3ffffff;
+        let r1 = u32::from_le_bytes(key[3..7].try_into().expect("infalliable conversion")) >> 2
+            & 0x03ffff03;
+        let r2 = u32::from_le_bytes(key[6..10].try_into().expect("infalliable conversion")) >> 4
+            & 0x03ffc0ff;
+        let r3 = u32::from_le_bytes(key[9..13].try_into().expect("infalliable conversion")) >> 6
+            & 0x03f03fff;
+        let r4 = u32::from_le_bytes(key[12..16].try_into().expect("infalliable conversion")) >> 8
+            & 0x000fffff;
+        let r = [r0, r1, r2, r3, r4];
+        let s0 = u32::from_le_bytes(key[16..20].try_into().expect("infalliable conversion"));
+        let s1 = u32::from_le_bytes(key[20..24].try_into().expect("infalliable conversion"));
+        let s2 = u32::from_le_bytes(key[24..28].try_into().expect("infalliable conversion"));
+        let s3 = u32::from_le_bytes(key[28..32].try_into().expect("infalliable conversion"));
+        let s = [s0, s1, s2, s3];
+
+        Poly1305 { r, s, acc: [0; 5], leftovers: [0u8; 16], leftovers_len: 0 }
+    }
+
+    /// Add message to be authenticated, can be called multiple times before creating tag.
+    pub fn add(&mut self, message: &[u8]) {
+        // Deal with previous leftovers if message is long enough.
+        let fill = if self.leftovers_len > 0 && (self.leftovers_len + message.len() >= 16) {
+            16 - self.leftovers_len
```

Comment:
In 74306fbde17bbbc09e44c388af56a081d8fb7dbf:

Why the condition that `self.leftovers_len > 0`? Tests all pass without it, and while it does change the control flow a little bit, I believe exactly the same amount of done without it. So better to simplify the condition.

Reply:
Yea, I think my mental model was to only run this first, one-off-y cycle iff there are previous leftovers and the message can fill the rest of the leftovers buffer. But dropping the previous leftovers requirement just means the first cycle just happens earlier on the message (assuming it is long enough). I'll drop the req and leave a comment describing this since it might be a little confusing.

Reply:
Thanks! I think it's easier to follow without the extra condition, which makes the whole `if` statement hard to model.